feat: host functions from NEP-616

jakmeier opened the pull request:
Add a bunch of new supporting host functions.

- current_contract_code
- promise_result_length
- refund_to_account_id
- storage_config_byte_cost
- storage_config_num_bytes_account
- storage_config_num_extra_bytes_record

All of these are to support the main feature or NEP-616 and described in the
same proposal.

jakmeier commented:
This is the last feature PR for NEP-616 and replaces the draft PR submitted for early review over at https://github.com/jakmeier/nearcore/pull/5

Before stabilising it, I also want to add integration tests with the test-loop setup. I'm still working on them but should have them ready today. My goal is to at least touch all the added host functions in those integration tests.

Reviewers, if you think this PR here could use unit-style tests, let me know what kind of tests would make sense and I will add them to this PR.

nagisa commented on runtime/near-vm-runner/src/logic/logic.rs:
```diff
@@ -725,6 +726,34 @@ impl<'a> VMLogic<'a> {
         )
     }
 
+    /// Saves the account id that was set for the current receipt by its
+    /// predecessor via [`Self::promise_set_refund_to()`], or
+    /// [`Self::predecessor_account_id()`] otherwise, into the register.
```
```suggestion
    /// Populates a register with the ID of an account which would receive a refund.
    ///
    /// This is the ID of an account set for the current receipt by its
    /// predecessor via [`Self::promise_set_refund_to()`], or
    /// [`Self::predecessor_account_id()`] otherwise.
```

nagisa commented on runtime/near-vm-runner/src/logic/logic.rs:
```diff
@@ -3187,6 +3216,45 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
+    /// # Cost
+    ///
+    /// `base` - the base cost for a simple host function call
+    pub fn promise_result_length(&mut self, result_idx: u64) -> Result<u64> {
```
What is the use-case for this host function, really? The description here says its about avoiding gas exhaustion, but what else would a contract do if the result is too long?

I find it somewhat not satisfying that this is a mechanism to avoid the problem of "data too long" for just one host function out of many that interact with registers. I feel it might be worth exploring if there is an opportunity for a size limiting mechanism that could be applied generally across everything dealing with registers.

nagisa commented on runtime/near-vm-runner/src/logic/logic.rs:
```diff
@@ -3724,6 +3845,45 @@ bls12381_p2_decompress_base + bls12381_p2_decompress_element * num_elements`
+    ///
+    /// `base` - the base cost for a simple host function call
+    /// `write_memory_base` + 16 * `write_memory_byte` - the cost of writing a `u128` to guest memory
+    pub fn storage_config_byte_cost(&mut self, balance_ptr: u64) -> Result<()> {
```
Similarly to the `promise_return_length` it might be quite unsatisfying if we end up needing to eventually add a host function for every single different configuration variable/fee that somebody might want to calculate with in their contract.

Could this be a a single function that takes the description of the fee to write into memory via an argument?

jakmeier commented:
> would be nice to also have unit tests for those, similar to #14364

added tests:

- 3 new tests in `runtime/near-vm-runner/src/logic/tests/storage_config.rs` covering
  - `storage_config_byte_cost`
  - `storage_config_num_bytes_account`
  - `storage_config_num_extra_bytes_record`
- 4 new tests in `runtime/near-vm-runner/src/logic/tests/miscs.rs` covering
  - `current_contract_code` (all 4 variations of code on an account)
- updated existing test in `runtime/near-vm-runner/src/logic/tests/promises.rs` to cover
    - `promise_result_length`
- 1 new test in `runtime/near-vm-runner/src/logic/tests/context.rs` convering 
    - `refund_to_account_id`

I think that's all :)
